### PR TITLE
ci(release): add rebase step before versioning to prevent push race

### PR DESCRIPTION
**Pull request by charlesmulder**
### Description
  
  Added rebase steps and deferred npm publish in release action to prevent push failures and orphaned npm versions.

  **Problem 1 - Push failures:** Release jobs check out `github.sha` (triggering commit), not main tip. When a PR merges to main while a release job is running, `git push --follow-tags` fails with non-fast-forward error.

  **Problem 2 - Tag resolution:** `verify-versions.sh` and `nx release` use `git tag --merged HEAD` to resolve current version. Stale checkout can't see tags from prior queued releases → false mismatches.

  **Problem 3 - Orphaned npm versions:** `nx release --yes` publishes to npm before `git push`. If push fails, npm has new versions but git doesn't → manual recovery required.

  **Solution:**
  1. Rebase before verify-versions (fast-forward, brings in latest tags so `--merged HEAD` sees them)
  2. `nx release --skip-publish` (defer npm publish)
  3. Rebase before push (replays version commit if main moved during release)
  4. `nx release publish` (only after push succeeds)

  **Already present:** Concurrency group with `cancel-in-progress: false` serializes release jobs.

  ---

  CI validation: Merge and observe next automated release. Check that:
  - Release job completes successfully
  - Version commits/tags pushed to main
  - Packages published to npm registry
  - No "non-fast-forward" errors in logs
  - No orphaned npm versions if push fails

  ---
  Anything reviewers should know?

  - Two rebase calls: first is always fast-forward (no local commits), second replays only the version commit
  - `--skip-publish` prevents publish in main `nx release` invocation; standalone `nx release publish` handles publish after push
  - If push fails (edge case: commit lands between rebase and push), re-running is safe — no npm versions published yet
  - Comments explain why each rebase is needed

  ---
  Checklist

  - [x] Tests: N/A (CI infrastructure change, validated by workflow execution)
  - [x] API: N/A (no API changes)
  - [x] Migrations: N/A (no schema changes)
  - [x] Dependencies: N/A (no dependency changes)
  - [x] Security: N/A (read-only git operations in trusted CI context)

  AI disclosure

  Assisted by: Claude Code

### CODE REVIEW

**Review by tahmidefaz — DISMISSED** [review]

**Review by justinorringer — DISMISSED** [review]

File: .github/actions/release/action.yaml [hunk 1]
```diff
@@ -32,6 +32,15 @@ runs:
...
         git config user.signingkey $(gpg --list-secret-keys --with-colons "${{ inputs.gh_email }}" | awk -F: '/^sec:/ {print $5; exit}')
+    - name: Rebase onto latest main
+      shell: bash
+      # nx release creates a version commit (package.json/CHANGELOG.md/package-lock.json
+      # bumps) and pushes it to master. Checkout is pinned to github.sha, so if a prior
+      # release already pushed its version commit, this checkout won't have it. Rebase
+      # incorporates those commits; without it, git push --follow-tags fails non-fast-forward.
+      run: |
+        git fetch origin main
+        git rebase origin/main
```
Comment (catastrophe-brandon) [inline .github/actions/release/action.yaml:43]: Won't a rebase operation remove the associated commit signature?

Comment (charlesmulder) [reply .github/actions/release/action.yaml:43]: Thanks for the review @catastrophe-brandon. Your review prompted quite a few changes. 

The commit that triggers the release job is checked out in CI. In most cases, it's a merge commit signed by Github. When `nx release` runs, `nx` creates a new commit containing updated `package.json` files and package changelogs. 

I've moved the rebase to just before pushing the commit `nx` created. This is to avoid a scenario where another commit lands in master, while the release step is running, which will cause the push to fail. 

Rebasing at this point will only resign the commit by `nx` with the `nacho-bot` signature, which is already the case anyway. I tested on local and can confirm other commits retain original signatures.

File: .github/actions/release/action.yaml [hunk 2]
```diff
@@ -58,5 +58,8 @@ runs:
       # HUSKY=0 disables pre-push hooks in CI — tests are run separately by the CI pipeline.
       env:
         HUSKY: '0'
-      run: git push --follow-tags
+      run: |
+        git fetch origin main
+        git rebase origin/main
+        git push --follow-tags
```
Comment (petrsimon) [inline .github/actions/release/action.yaml:78]: Shouldn't it be rather `git push --force-with-lease --follow-tags`?

Comment (charlesmulder) [reply .github/actions/release/action.yaml:78]: Thanks for the review @petrsimon. Looks like it should be okay. 

After rebase, the version commit sits directly on top of the fetched origin/main — the push is always fast-forward. --force-with-lease is for replacing already-published history; here the version commit was just created locally and hasn't been pushed yet. If a commit lands between rebase and push (non-fast-forward), we want the push to fail so we can re-run safely — no npm versions were published.

**Review by justinorringer — DISMISSED** [review]

File: .github/scripts/verify-versions.sh [hunk 3]
```diff
@@ -54,13 +57,25 @@ for dir in "$REPO_ROOT"/packages/*/; do
...
+    else
+      npm_ver="REGISTRY_ERROR"
+    fi
+  fi
 
   # --- Compare ---
   if [[ "$disk_ver" == "$tag_ver" && "$disk_ver" == "$npm_ver" ]]; then
     status="OK"
     marker="✓"
+  elif [[ "$npm_ver" == "REGISTRY_ERROR" ]]; then
```
Comment (catastrophe-brandon) [inline .github/scripts/verify-versions.sh:73]: I'm wondering what happens if $npm_ver is "NOT_ON_NPM" here? Do we need another elif?

Comment (charlesmulder) [reply .github/scripts/verify-versions.sh:73]: Thanks for the review @catastrophe-brandon. I am not sure I understand the question? Execution will fall through to the next `elif` block or end up in the `else` block. `NOT_ON_NPM` means the package doesn't exist on npm. `REGISTRY_ERROR` means unable to reach the registry. Otherwise, `$npm_ver` should be a version number.

Comment (catastrophe-brandon) [reply .github/scripts/verify-versions.sh:73]: I think Claude can explain better than I can:

```
The problem: When $npm_ver is set to "NOT_ON_NPM", this condition will only pass if $disk_ver also equals the literal string "NOT_ON_NPM",
  which should never happen (disk version should be a semver like "1.2.3").

  What actually happens:
  1. Line 70 fails (as intended) because disk version won't match the string "NOT_ON_NPM"
  2. Control flow falls through to line 73 (registry error check) - doesn't match
  3. Control flow falls through to line 79 (new package check) - this catches the case where tag_ver == "NONE" && npm_ver == "NOT_ON_NPM"
  4. If tag exists but npm doesn't, we skip to line 86 (MISMATCH)

  The issue: If a package has a git tag (e.g., previously released) but is NOT_ON_NPM (which shouldn't happen in normal flow), it's treated
  as a generic MISMATCH without explaining that the package isn't on npm. The error message on line 89 would show npm=NOT_ON_NPM but doesn't
  explicitly say "this package should be on npm but isn't."

  Recommendation: Add a specific check for the case where we have a tag but NOT_ON_NPM:

  elif [[ "$tag_ver" != "NONE" && "$npm_ver" == "NOT_ON_NPM" ]]; then
    status="PUBLISH_FAILED"
    marker="✗"
    mismatch=1
    echo "::error::${pkg}: package has git tag but is not on npm (disk=${disk_ver}, tag=${tag_ver}). Publish likely failed."

  This would make it clearer when a previously-tagged release never made it to npm.
```

Comment (catastrophe-brandon) [reply .github/scripts/verify-versions.sh:73]: This isn't something that should block the PR approval, so I'd be fine with tackling it in a future PR.

**Review by catastrophe-brandon — APPROVED** [review]

**Review by karelhala — APPROVED** [review]
